---
title: "label transfer"
output: html_document
---

```{r setup, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(rtracklayer)
library(ggplot2)
```

## Input files and paths
```{r}
# Integrated ATAC object produced previously ("intg30" object)

atac_int <- readRDS("/path.to/intg30.rds")

# Labeled RNA produced previously

rna  <-readRDS("/path.to/out.rds")

# GTF used for gene annotations (same reference as ATAC processing)

gtf <- "/path.to/cellranger/refdata-cellranger-arc-GRCh38-2020-A-2.0.0/genes/genes.gtf.gz"


```

## Prepare ATAC object for label transfer
Add gene annotations to ATAC object
```{r}
DefaultAssay(atac_int) <- "peaks"

# GTF used for gene annotations (same reference as ATAC processing)
gtf <- rtracklayer::import("/path.to/cellranger/refdata-cellranger-arc-GRCh38-2020-A-2.0.0/genes/genes.gtf.gz")
genes <- gtf[gtf$type == "gene"]

# Ensure required metadata columns exist

mcols(genes)$gene_id <- as.character(mcols(genes)$gene_id)
if (!"gene_name" %in% colnames(mcols(genes))) {
mcols(genes)$gene_name <- mcols(genes)$gene_id
}
if (!"gene_biotype" %in% colnames(mcols(genes))) {


  if ("gene_type" %in% colnames(mcols(genes))) {
mcols(genes)$gene_biotype <- as.character(mcols(genes)$gene_type)
} else {
mcols(genes)$gene_biotype <- NA_character_
}
}

# Harmonize chromosome naming

seqlevels(genes) <- paste0("chr", gsub("^chr", "", seqlevels(genes)))
genome(genes) <- "hg38"

Annotation(atac_int) <- genes
is.null(Annotation(atac_int))

```

Compute gene activity and add as assay
```{r}
ga <- GeneActivity(atac_int)

atac_int[["ACTIVITY"]] <- CreateAssayObject(counts = ga)
Assays(atac_int)

```


Find transfer anchors
```{r}
DefaultAssay(atac_int) <- "ACTIVITY"
atac_int <- NormalizeData(atac_int, verbose = FALSE)
atac_int <- FindVariableFeatures(atac_int, verbose = FALSE)
atac_int <- ScaleData(atac_int, verbose = FALSE)

DefaultAssay(atac_int)
length(VariableFeatures(atac_int))

```


## Transfer cell type labels
```{r}
pred <- TransferData(
anchorset = transfer.anchors,
refdata = rna$celltype,
weight.reduction = atac_int[["integrated_lsi"]],
dims = 2:30
)

atac_int <- AddMetaData(atac_int, pred)

```


## Visualise transferred lables on the integrated UMAP
```{r}
DimPlot(
atac_int,
reduction = "umap_integrated",
group.by = "predicted.id",
label = TRUE,
repel = TRUE
) + NoLegend()

```

```{r}
# save object
saveRDS(atac_int, "labelled_atac.rds")
```

